Replace the Dataout::first_cell/next_cell() mechanism

bangerth opened the pull request:
Instead, use function objects that can be used with lambda functions. This addresses the idea mentioned in https://github.com/dealii/dealii/issues/9099#issuecomment-571775767

The current approach is backward compatible in that the new function objects by default just call the old function. As a consequence, all of the patches currently floating around (#9100, #9096) should continue to work -- but could be streamlined a bit by just calling `DataOut::set_cell_selection()` in their constructors instead of implementing several more functions.

/rebuild

masterleinad commented on doc/news/changes/minor/20200111Bangerth:
```diff
@@ -0,0 +1,6 @@
+New: The mechanism implemented by Dataout via the `first_cell()` and
```
I am pretty sure that you need `DataOut::first_cell()` and `DataOut::next_cell()` for `doxygen` to pick it up.

bangerth replied:
I specifically didn't want doxygen to pick these names up because I intend to remove these functions :-)

masterleinad replied:
I see. But I guess you still want 
```suggestion
New: The mechanism implemented by DataOut via the `first_cell()` and
```

tjhei commented:
> I find that an implementation with `FilteredIterator` as suggested by @tjhei would be cleaner.

FilteredIterator also turns out to be a bit awkward to use in practice as the predicate is part of the type of begin and end.

Can we change set_cell_selection to take a single argument that allows passing ``IteratorRange< FilteredIterator< BaseIterator > >``
so we can use
``filter_iterators(triangulation.active_cell_iterators(),IteratorFilters::LocallyOwnedCell())``
and
``triangulation.cell_iterators_on_level(lvl)``
etc.? See https://www.dealii.org/developer/doxygen/deal.II/group__CPP11.html#ga3948dd66d9e66ac55d89dc8f75fa4855

bangerth commented on include/deal.II/numerics/data_out.h:
```diff
@@ -316,42 +313,117 @@ class DataOut : public DataOut_DoFData<DoFHandlerType,
+   * or to completely change *where* output is produced (e.g., to produce
+   * output on non-active cells of a multigrid hierarchy or if the finest
+   * level of a mesh is so fine that generating graphical output would lead
+   * to an overwhelming amount of data.
```
Let's not document things that aren't possible *yet* :-)

bangerth commented on include/deal.II/numerics/data_out.h:
```diff
@@ -316,42 +313,117 @@ class DataOut : public DataOut_DoFData<DoFHandlerType,
+   * @param[in] A function object that takes as argument the last cell
+   *   on which output was generated and that should return the next
+   *   cell on which output should be generated. (In practice, the
+   *   function object may be triggered *before* any processing
```
I've removed the sentence. I meant to say that *all* of the walking over cells might happen before the first output is generated. But that's really immaterial to the user to know.

bangerth commented on include/deal.II/numerics/data_out.h:
```diff
@@ -316,42 +313,117 @@ class DataOut : public DataOut_DoFData<DoFHandlerType,
+   *   on which output was generated and that should return the next
+   *   cell on which output should be generated. (In practice, the
+   *   function object may be triggered *before* any processing
+   *   actually happens, but you get the idea.)
```
I documented both. Since the triangulation is so easy to obtain, I think there's no point in passing it explicitly (and most implementations wouldn't need it because they just do something like `++cell`.

bangerth replied:
Actually, I change my mind.

bangerth commented on source/numerics/data_out.cc:
```diff
@@ -65,6 +65,24 @@ namespace internal
+  // functions created here.
+  set_cell_selection(
+    [this](const Triangulation<dim, spacedim> &) {
+      return this->first_locally_owned_cell();
```
First, no it doesn't trigger a deprecation warning: We disable these while compiling the library itself.

Second: For backward compatibility, this is exactly the behavior we want to implement: To call the existing functions. Only when we remove the deprecated functions should we inline their implementations into the lambda function here.

bangerth commented:
I've addressed the comments made before, and pushed another commit that implements a second version of the function. What do you think of the latter?

I'll caution that the new version does not select locally owned cells. I leave this to the provided filter -- or should I just filter on that as well?

tjhei commented:
> I've addressed the comments made before, and pushed another commit that implements a second version of the function. What do you think of the latter?

I think usability should be much better with the filtered iterators compared to writing the lambda functions. :+1:  It might not be a bad idea to keep both around for more flexibility.

> I'll caution that the new version does not select locally owned cells. I leave this to the provided filter -- or should I just filter on that as well?

1. Your documentation currently says that it filters for locally owned (even though it doesn't). I can think of one case where I would not want to filter out locally_owned_cells: to visualize what part of the mesh each processor sees, including ghost cells. Maybe this is not too important as we have done it with GridOut in the past. It gets more complicated to do the right thing if you allow non-active cells, though (owned vs owned on level). Maybe let the user deal with it?
2. You are currently using ``begin_active()`` but ``cell_iterator``. Does this mean you cover active and non-active cells but only starting from begin_active() right now? I would want this function to work on all cells (begin() to end()) by default.

bangerth commented:
Good points. I've addressed them both in the second to last commit. (I've changed the description of the function, and I've made the iterator start at `.begin()` instead of `.begin_active()` -- that was a mistake.)

tjhei commented:
I like it!
To make things testable, can we keep the first/next_cell() public functions and make them call the lambda functions. Then we can have simple tests like in  #9278 to check if the API works as intended (select a subset of cells, non-active ones, etc).

bangerth commented:
Well, I specifically wanted to get away from the virtual functions because I think that's a rather clunky interface. Right now, the lambdas by default call the virtual functions. But the test will be easy to adapt: You can keep the (virtual or not) function in your class and in the constructor you just call the `set_cell_selection()` function with two lambda functions that just call the functions in your class.

With this patch, we now have two interfaces -- one via the virtual functions (which the patch deprecates) and one via the function objects. I'd rather not keep both of these interfaces around indefinitely.

tjhei commented:
> Well, I specifically wanted to get away from the virtual functions because I think that's a rather clunky interface.

Alright. I don't want to keep them as a way to customize behavior by deriving, I suggested to have something to unit test the implementation of set_cell_selection().

> With this patch, we now have two interfaces -- one via the virtual functions (which the patch deprecates) and one via the function objects. I'd rather not keep both of these interfaces around indefinitely.

If the deprecated functions are removed, I don't know of a way to test that what you implemented in this PR actually works (the FilteredIterator logic for example). Would you mind writing some tests that show that what you implemented works or suggest how you would test it?

bangerth commented:
Sure, I just wanted to see whether everyone likes the design first. Two tests for the two functions are in the last commit now! (Turns out that there was also a bug in my code for the second `set_cell_selection()` before -- good to have checked!)

tjhei commented on tests/data_out/data_out_08_set_cell_selection_01.cc:
```diff
@@ -0,0 +1,144 @@
+
+  // this should skip the first cell
+  typename Triangulation<dim>::active_cell_iterator it = tria.begin_active();
+  //++it;
```
please remove

tjhei commented on tests/data_out/data_out_08_set_cell_selection_01.cc:
```diff
@@ -0,0 +1,144 @@
+  DoFHandler<dim> dof_handler(tria);
+  dof_handler.distribute_dofs(fe);
+
+  //  DataOut<dim> data_out;
```
please remove

tjhei commented on tests/data_out/data_out_08_set_cell_selection_02.cc:
```diff
@@ -0,0 +1,128 @@
+
+  // this should skip the first cell
+  typename Triangulation<dim>::active_cell_iterator it = tria.begin_active();
+  //++it;
```
please remove

tjhei commented on tests/data_out/data_out_08_set_cell_selection_02.cc:
```diff
@@ -0,0 +1,128 @@
+  DoFHandler<dim> dof_handler(tria);
+  dof_handler.distribute_dofs(fe);
+
+  //  DataOut<dim> data_out;
```
please remove

tjhei commented:
can you please rebase to master?

bangerth commented:
Fixed. I thought about this some more, and it turns out that one can convert lambda functions to `FilteredIterator` objects. Take a look at the last commit -- I think you're really going to like this :-)

peterrum commented:
>  I think you're really going to like this :-)

Yup ;)

tjhei commented:
awesome!
